include/boost/geometry/algorithms/detail/envelope/multipoint.hpp

```diff
@@ -0,0 +1,362 @@
+
+#include <boost/geometry/util/convert_on_spheroid.hpp>
+#include <boost/geometry/util/math.hpp>
+#include <boost/geometry/util/normalize_spheroidal_coordinates.hpp>
+
+#include <boost/geometry/algorithms/assign.hpp>
+#include <boost/geometry/algorithms/convert.hpp>
+
+#include <boost/geometry/algorithms/detail/envelope/range.hpp>
+#include <boost/geometry/algorithms/dispatch/envelope.hpp>
+
+
+namespace boost { namespace geometry
+{
+
+#ifndef DOXYGEN_NO_DETAIL
+namespace detail { namespace envelope
+{
+
+
+class envelope_multipoint_on_spheroid
+{
+private:
+    template <typename T>
+    struct lon_lat_pair
+        : std::pair<T, T>
+    {
+        lon_lat_pair(T const& first, T const& second)
+            : std::pair<T, T>(first, second)
+        {}
+    };
+
+    struct longitude_less
+    {
+        template <typename Pair>
+        inline bool operator()(Pair const& pair1, Pair const& pair2) const
+        {
+            return math::smaller(pair1.first, pair2.first);
+        }
+    };
+
+    struct latitude_less
+    {
+        template <typename Pair>
+        inline bool operator()(Pair const& pair1, Pair const& pair2) const
+        {
+            return math::smaller(pair1.second, pair2.second);
+        }
+    };
+
+
+    template <typename Constants, typename MultiPoint, typename OutputIterator>
+    static inline void analyze_point_coordinates(MultiPoint const& multipoint,
+                                                 bool& has_south_pole,
+                                                 bool& has_north_pole,
+                                                 OutputIterator oit)
+    {
+        typedef typename boost::range_iterator
+            <
+                MultiPoint const
+            >::type iterator_type;
+        typedef typename coordinate_type<MultiPoint>::type coordinate_type;
+
+        // analyze point coordinates:
+        // (1) normalize point coordinates
+        // (2) check if any point is the north or the south pole
+        // (3) put all non-pole points in a container
+        //
+        // notice that at this point in the algorithm, we have at
+        // least two points on the spheroid
+        has_south_pole = has_north_pole = false;
+
+        for (iterator_type it = boost::begin(multipoint);
+             it != boost::end(multipoint);
+             ++it)
+        {
+            coordinate_type lon = geometry::get<0>(*it);
+            coordinate_type lat = geometry::get<1>(*it);
+
+            math::normalize_spheroidal_coordinates
+                <
+                    typename coordinate_system<MultiPoint>::type::units,
+                    coordinate_type
+                >(lon, lat);
+
+            if (math::equals(lat, Constants::min_latitude()))
+            {
+                has_south_pole = true;
+            }
+            else if (math::equals(lat, Constants::max_latitude()))
+            {
+                has_north_pole = true;
+            }
+            else
```

Comment:
Yes, indeed using the original coordinates wherever possible seems to be a better idea.